### PR TITLE
Add ir_to_markdown.py utility script.

### PR DESCRIPTION
**Pull request by ScottTodd**
Progress on https://github.com/google/iree/issues/5958, and could be generally useful.

Sample output: https://gist.github.com/ScottTodd/a9de0470eb4cead7a84712c18bff9426.

Note: I get 73 MB from compiling `bert_encoder_unrolled_fake_weights.mlir`, which fails to upload to a gist :P

### CODE REVIEW

**Comment by benvanik** [comment]
If this works with iree-opt then having partial pipelines may make it more useful on larger models - for example, running just the -iree-flow-transformation-pipeline, -iree-hal-transformation-pipeline, etc - at least partitions it by major level of the IR for hopefully <50MB files :P

**Comment by benvanik** [comment]
Could also have it elide unchanged IR - if a canonicalizer pass runs and nothing changes, just emit "no change" instead of all the IR again. Would also improve readability when scanning to see where passes did meaningful work.

**Comment by ScottTodd** [comment]
> Could also have it elide unchanged IR - if a canonicalizer pass runs and nothing changes, just emit "no change" instead of all the IR again. Would also improve readability when scanning to see where passes did meaningful work.

There is `print-ir-after-change`: https://mlir.llvm.org/docs/PassManagement/#ir-printing.

I have two goals here:

* The linked issue, which would involve automated publishing of (perhaps elided) IR dumps
* Displaying IR in a pretty format for use with https://scotttodd.github.io/iree-llvm-sandbox/web-tools/pipeline-visualizer/ (and so I have something to talk to with https://llvm.discourse.group/t/data-visualizations-and-correlating-data-between-passinstrumentations/3870)

Both would benefit from pruning passes that do not change the IR and from choosing partial pipelines.

**Comment by ScottTodd** [comment]
> If this works with iree-opt then having partial pipelines may make it more useful on larger models - for example, running just the -iree-flow-transformation-pipeline, -iree-hal-transformation-pipeline, etc - at least partitions it by major level of the IR for hopefully <50MB files :P

`iree-opt` should work, though I've started using `iree-translate` for this sort of printing for real models, since I haven't been able to get `-iree-input-type=mhlo` to work with `iree-opt` :/

File: scripts/ir_to_markdown.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+                      nargs='?',
+                      metavar="<input_file_path>",
+                      help='Input IR dump (.mlir from -print-ir-after-all)')
+  parser.add_argument('-o,',
+                      '--output',
+                      type=str,
+                      required=True,
+                      metavar="<output>",
+                      help='Output file path (e.g. translation_ir.md)')
+  # TODO(scotttodd): flag for input IR path
```
Comment (GMNGeoffrey) [inline scripts/ir_to_markdown.py:50]: Seems like we have that?

File: scripts/ir_to_markdown.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+  # TODO(scotttodd): flag for command line used for compilation
+  # TODO(scotttodd): flag for markdown flavor (mkdocs, github, etc.)
+  # TODO(scotttodd): flag for diff view (correlate IR before and IR after)?
```
Comment (ScottTodd) [inline scripts/ir_to_markdown.py:53]: BTW, I prototyped this in a gist: https://gist.github.com/ScottTodd/191edcc56bbc7998ac0c2704c9793863

Outside of GitHub flavored markdown we have more options for how to display diffs.

**Review by GMNGeoffrey — APPROVED** [review]

File: scripts/ir_to_markdown.py (hunk 1 above)
Comment (ScottTodd) [reply scripts/ir_to_markdown.py:50]: Ah, I meant for the "original" IR (`$PWD/iree/samples/models/simple_abs.mlir` in this example, not `/tmp/simple_abs_vmvx_pipeline.mlir`). Like how on https://github.com/google/iree/blob/99c6ce409783eaffb37dd46f3f4a6ef761556f99/docs/ir_examples/mhlo-dot-vmla.md we had a little introduction, the command used, the input IR, then the IR conversion.

I'll reword this and the sample comment.

**Comment by ScottTodd** [comment]
Adding `-print-ir-after-change` drops bert from 74MB to 58MB. Still too large to upload as a .md file to a gist though.

**Pull request merged**